File: etrike_protocol/codecs/typescript/seb.ts
import { compute, verify } from "../../profiles/xor8_ff_v1";
import { type CanFrame, type CodecStatus, type DecodeResult, type EncodeResult, frame, validateFrame } from "./types";

export const SEB_BUS = "low";
export const SEB_DLC = 8;
export const SEB_COMMAND_ID = 0x7b9;
export const SEB_STATUS_ID = 0x721;
export const SEB_ERROR_INFO_ID = 0x731;
export const SEB_VERSION_ID = 0x741;
export const SEB_TEST_ID = 0x6fb;

function validate(input: CanFrame, id: number, checksum = false): CodecStatus {
  const status = validateFrame(input, { bus: SEB_BUS, id, frameFormat: "standard", dlc: SEB_DLC });
  if (status !== "ok") return status;
  if (checksum && !verify(input.data.subarray(0, 7), input.data[7])) return "checksum_mismatch";
  return "ok";
}

function integer(value: unknown, fallback: number): number | undefined {
  const selected = value ?? fallback;
  return typeof selected === "number" && Number.isInteger(selected) ? selected : undefined;
}

export function encodeSebCommand(values: Readonly<Record<string, unknown>>): EncodeResult {
  if (
    (values.alignment_enable !== undefined && typeof values.alignment_enable !== "boolean") ||
    (values.control_enable !== undefined && typeof values.control_enable !== "boolean") ||
    (values.auto_brake !== undefined && typeof values.auto_brake !== "boolean")
  ) return ["value_out_of_range"];
  const mode = integer(values.control_mode, 0);
  const stroke = integer(values.stroke_request_raw, 600);
  const pressure = integer(values.pressure_request_raw, 0);
  const counter = integer(values.rolling_counter, 0);
  if (mode === undefined || ![0, 1].includes(mode)) return ["invalid_enum"];
  if (stroke === undefined || pressure === undefined || counter === undefined) return ["value_out_of_range"];
  if (stroke < 0 || stroke > 0xffff || pressure < 0 || pressure > 100 || counter < 0 || counter > 15) {
    return ["value_out_of_range"];
  }
  const payload = new Uint8Array(SEB_DLC);
  payload[0] =
    (values.alignment_enable === true ? 1 : 0) |
    (values.control_enable === true ? 2 : 0) |
    (mode << 2) |
    (values.auto_brake === true ? 8 : 0);
  payload[2] = stroke & 0xff;
  payload[3] = mode === 0 ? stroke >> 8 : pressure;
  payload[6] = 0x03 | (counter << 4);
  payload[7] = compute(payload.subarray(0, 7));
  return ["ok", frame(SEB_BUS, SEB_COMMAND_ID, "standard", payload)];
}

export function decodeSebCommand(input: CanFrame): DecodeResult<Record<string, unknown>> {
  const status = validate(input, SEB_COMMAND_ID, true);
  if (status !== "ok") return [status];
  if ((input.data[6] & 0x03) !== 0x03) return ["constant_mismatch"];
  const mode = (input.data[0] & 4) !== 0 ? 1 : 0;
  const pressure = mode === 1 ? input.data[3] : 0;
  if (pressure > 100) return ["value_out_of_range"];
  return ["ok", {
    alignment_enable: (input.data[0] & 1) !== 0,
    control_enable: (input.data[0] & 2) !== 0,
    control_mode: mode,
    auto_brake: (input.data[0] & 8) !== 0,
    stroke_request_raw: mode === 0 ? input.data[2] | (input.data[3] << 8) : input.data[2],
    pressure_request_raw: pressure,
    rolling_counter: input.data[6] >> 4,
  }];
}

export function decodeSebStatus(input: CanFrame): DecodeResult<Record<string, unknown>> {
  const status = validate(input, SEB_STATUS_ID, true);
  if (status !== "ok") return [status];
  const view = new DataView(input.data.buffer, input.data.byteOffset, input.data.byteLength);
  return ["ok", {
    status_byte: input.data[0],
    alignment_status: (input.data[0] & 1) !== 0,
    control_enabled: (input.data[0] & 2) !== 0,
    control_mode: (input.data[0] >> 2) & 3,
    auto_brake_status: (input.data[0] & 0x10) !== 0,
    error_status: (input.data[0] >> 6) & 3,
    stroke_value_raw: view.getUint16(2, true),
    pressure_value_raw: input.data[3],
    angle_value_raw: view.getInt16(5, true),
    rolling_counter_enabled: (input.data[6] & 1) !== 0,
    checksum_enabled: (input.data[6] & 2) !== 0,
    rolling_counter: input.data[6] >> 4,
  }];
}

export function decodeSebErrorInfo(input: CanFrame): DecodeResult<{ raw: Uint8Array }> {
  const status = validate(input, SEB_ERROR_INFO_ID);
  return status === "ok" ? ["ok", { raw: input.data.slice() }] : [status];
}

export function decodeSebVersion(input: CanFrame): DecodeResult<Record<string, unknown>> {
  const status = validate(input, SEB_VERSION_ID);
  return status === "ok"
    ? ["ok", { software_raw: input.data[0], hardware_raw: input.data[1], raw: input.data.slice() }]
    : [status];
}

export function decodeSebTest(input: CanFrame): DecodeResult<Record<string, number>> {
  const status = validate(input, SEB_TEST_ID);
  if (status !== "ok") return [status];
  const view = new DataView(input.data.buffer, input.data.byteOffset, input.data.byteLength);
  return ["ok", {
    motor_current_raw: view.getInt16(1, true),
    ecu_temperature_raw: view.getUint16(3, true),
    supply_voltage_raw: view.getUint16(5, true),
  }];
}

File: etrike_protocol/codecs/python/seb.py
from __future__ import annotations

from collections.abc import Mapping

from protocol.profiles.xor8_ff_v1 import compute, verify

from .types import CodecStatus, Frame, validate_frame

BUS = "low"
DLC = 8
COMMAND_ID = 0x7B9
STATUS_ID = 0x721
ERROR_INFO_ID = 0x731
VERSION_ID = 0x741
TEST_ID = 0x6FB


def _validate(frame: Frame, can_id: int, *, checksum: bool = False) -> CodecStatus:
    status = validate_frame(frame, bus=BUS, can_id=can_id, frame_format="standard", dlc=DLC)
    if status != "ok":
        return status
    if checksum and not verify(frame.data[:7], frame.data[7]):
        return "checksum_mismatch"
    return "ok"


def _integer(value: object, default: int) -> int | None:
    selected = default if value is None else value
    return selected if isinstance(selected, int) and not isinstance(selected, bool) else None


def encode_command(values: Mapping[str, object]) -> tuple[CodecStatus, Frame | None]:
    alignment_enable = values.get("alignment_enable", False)
    control_enable = values.get("control_enable", False)
    auto_brake = values.get("auto_brake", False)
    mode = _integer(values.get("control_mode"), 0)
    stroke = _integer(values.get("stroke_request_raw"), 600)
    pressure = _integer(values.get("pressure_request_raw"), 0)
    counter = _integer(values.get("rolling_counter"), 0)
    if not all(isinstance(value, bool) for value in (alignment_enable, control_enable, auto_brake)):
        return "value_out_of_range", None
    if stroke is None or pressure is None or counter is None:
        return "value_out_of_range", None
    if mode not in (0, 1):
        return "invalid_enum", None
    if not 0 <= stroke <= 0xFFFF or not 0 <= pressure <= 100 or not 0 <= counter <= 15:
        return "value_out_of_range", None
    payload = bytearray(DLC)
    payload[0] = (
        int(alignment_enable)
        | (int(control_enable) << 1)
        | (mode << 2)
        | (int(auto_brake) << 3)
    )
    payload[2] = stroke & 0xFF
    payload[3] = stroke >> 8 if mode == 0 else pressure
    payload[6] = 0x03 | (counter << 4)
    payload[7] = compute(payload[:7])
    return "ok", Frame(BUS, COMMAND_ID, "standard", payload)


def decode_command(frame: Frame) -> tuple[CodecStatus, dict[str, object] | None]:
    status = _validate(frame, COMMAND_ID, checksum=True)
    if status != "ok":
        return status, None
    if frame.data[6] & 0x03 != 0x03:
        return "constant_mismatch", None
    mode = 1 if frame.data[0] & 0x04 else 0
    pressure = frame.data[3] if mode == 1 else 0
    if pressure > 100:
        return "value_out_of_range", None
    return "ok", {
        "alignment_enable": bool(frame.data[0] & 0x01),
        "control_enable": bool(frame.data[0] & 0x02),
        "control_mode": mode,
        "auto_brake": bool(frame.data[0] & 0x08),
        "stroke_request_raw": frame.data[2] | (frame.data[3] << 8) if mode == 0 else frame.data[2],
        "pressure_request_raw": pressure,
        "rolling_counter": frame.data[6] >> 4,
    }


def decode_status(frame: Frame) -> tuple[CodecStatus, dict[str, object] | None]:
    status = _validate(frame, STATUS_ID, checksum=True)
    if status != "ok":
        return status, None
    return "ok", {
        "status_byte": frame.data[0],
        "alignment_status": bool(frame.data[0] & 0x01),
        "control_enabled": bool(frame.data[0] & 0x02),
        "control_mode": (frame.data[0] >> 2) & 0x03,
        "auto_brake_status": bool(frame.data[0] & 0x10),
        "error_status": (frame.data[0] >> 6) & 0x03,
        "stroke_value_raw": int.from_bytes(frame.data[2:4], "little"),
        "pressure_value_raw": frame.data[3],
        "angle_value_raw": int.from_bytes(frame.data[5:7], "little", signed=True),
        "rolling_counter_enabled": bool(frame.data[6] & 0x01),
        "checksum_enabled": bool(frame.data[6] & 0x02),
        "rolling_counter": frame.data[6] >> 4,
    }


def decode_error_info(frame: Frame) -> tuple[CodecStatus, dict[str, bytes] | None]:
    status = _validate(frame, ERROR_INFO_ID)
    return (status, {"raw": frame.data} if status == "ok" else None)


def decode_version(frame: Frame) -> tuple[CodecStatus, dict[str, object] | None]:
    status = _validate(frame, VERSION_ID)
    if status != "ok":
        return status, None
    return "ok", {"software_raw": frame.data[0], "hardware_raw": frame.data[1], "raw": frame.data}


def decode_test(frame: Frame) -> tuple[CodecStatus, dict[str, int] | None]:
    status = _validate(frame, TEST_ID)
    if status != "ok":
        return status, None
    return "ok", {
        "motor_current_raw": int.from_bytes(frame.data[1:3], "little", signed=True),
        "ecu_temperature_raw": int.from_bytes(frame.data[3:5], "little"),
        "supply_voltage_raw": int.from_bytes(frame.data[5:7], "little"),
    }
